### ISSUE TITLE
Calculate time consuming for a request

### ISSUE THREAD

**Issue by Asilent** [Kong/kong#7089]
I want to calculate how long does a request need. I use plugin as follows:
```
function plugin:access(plugin_conf)
      kong.ctx.plugin.request_access_time = os.clock()


function plugin:log(plugin_conf)
     local elapse_time = (os.clock() - kong.ctx.plugin.request_access_time)*1000 --milliseconds
```

It returns a error result. The real time consuming is 300ms but I just get 20ms.

**Comment by fffonion** [comment]
Plugin handlers are not ran at the very beginning each phase, instead consider using data provided by Kong like
https://github.com/Kong/kong/blob/master/kong/pdk/log.lua#L736 by using
```lua
local elapse_time  = kong.log.serialize().latencies.kong
-- or
local ctx = ngx.ctx
local elapse_time = (ctx.KONG_PROXY_LATENCY or ctx.KONG_RESPONSE_LATENCY or 0) +
                 (ctx.KONG_RECEIVE_TIME or 0)
```

**Comment by bungle** [comment]
try using:

```lua
function plugin:access(plugin_conf)
      ngx.update_time()
      kong.ctx.plugin.request_access_time = ngx.now()
end

function plugin:log(plugin_conf)
     ngx.update_time()
     local elapse_time = (ngx.now() - kong.ctx.plugin.request_access_time) * 1000 --milliseconds
 end
 ```

**Comment by Asilent** [comment]
> try using:
> 
> 
> 
> ```lua
> 
> function plugin:access(plugin_conf)
> 
>       ngx.update_time()
> 
>       kong.ctx.plugin.request_access_time = ngx.now()
> 
> end
> 
> 
> 
> function plugin:log(plugin_conf)
> 
>      ngx.update_time()
> 
>      local elapse_time = (ngx.now() - kong.ctx.plugin.request_access_time) * 1000 --milliseconds
> 
>  end
> 
>  ```
> 
>  
> 
> But please note, it is not guaranteed when the log handler runs (so perhaps the better to collect time at last call to `body_filter`- but we already do it: https://github.com/Kong/kong/blob/master/kong/init.lua#L1197). Also there can be a lot of phases prior `access`.  You might want to use `ngx.var.request_time` instead:
> 
>  
> 
>  ```lua
> 
>  function plugin:log(plugin_conf)
> 
>      ngx.update_time() -- most probably not neccessary
> 
>      local elapse_time = (ngx.now() * 1000 - tonumber(ngx.var.request_time, 10))
> 
>  end
> 
>  ```

The time I get using the ngx.var.request_time parameter is a decimal number in seconds. This looks more like elapse_time. Why do I use ngx.now()*1000 to subtract it to get elapse_time.

**Comment by bungle** [comment]
> The time I get using the ngx.var.request_time parameter

Ok, then correct the code, simple as that. I thought it was milliseconds from docs:
http://nginx.org/en/docs/http/ngx_http_core_module.html#var_request_time 

But now that I read that again, isn't that what you want? e.g. `ngx.var.request_time` only? Or what do you mean by `how long does a request need`?

**Comment by bungle** [comment]
That is what our log serialize gives you by default:
https://github.com/Kong/kong/blob/master/kong/pdk/log.lua#L739

**Comment by Asilent** [comment]
> But now that I read that again, isn't that what you want? e.g. `ngx.var.request_time` (or `ngx.var.request_time * 1000`) only? Or what do you mean by `how long does a request need`?



Now I have realized that ngx.var.request_time will give what I want. Before, I tried to use os.clock - access_request_time to get that in log phase. But could you interpret this equation：

```
function plugin:log(plugin_conf)
    ngx.update_time() -- most probably not neccessary
    local elapse_time = (ngx.now() * 1000 - tonumber(ngx.var.request_time, 10))
end
```

I don't quite get what it refers to.

**Comment by bungle** [comment]
@Asilent,

Yes, I originally thought wrongly that `ngx.var.request_time` was a single value when the request hit the nginx, but it is always a calculated value of how much time it has taken at that point when it is read.

So here is a corrected version:
```
function plugin:log(plugin_conf)
    local elapse_time = ngx.var.request_time * 1000
end
```

**Comment by Asilent** [comment]
@bungle Thanks very much.
